diff -ur linux-2.4.0-orig/MAINTAINERS linux-2.4.0-test8/MAINTAINERS
--- linux-2.4.0-orig/MAINTAINERS	Thu Sep 14 22:39:51 2000
+++ linux-2.4.0-test8/MAINTAINERS	Sun Sep 17 10:01:49 2000
@@ -848,6 +848,13 @@
 W:	http://www.linuxtr.net
 S:	Maintained
 
+ONSTREAM SCSI TAPE DRIVER
+P:	Willem Riede
+M:	osst@riede.org
+L:	osst@linux1.onstream.nl
+L:	linux-scsi@vger.rutgers.edu
+S:	Maintained
+
 OPL3-SA2, SA3, and SAx DRIVER
 P:	Scott Murray
 M:	scott@spiteful.org
diff -ur linux-2.4.0-orig/drivers/scsi/Makefile linux-2.4.0-test8/drivers/scsi/Makefile
--- linux-2.4.0-orig/drivers/scsi/Makefile	Sun Aug  6 14:23:40 2000
+++ linux-2.4.0-test8/drivers/scsi/Makefile	Sun Sep 17 10:01:49 2000
@@ -30,6 +30,7 @@
 
 obj-$(CONFIG_SCSI)		+= scsi_mod.o
 obj-$(CONFIG_CHR_DEV_ST)	+= st.o
+obj-$(CONFIG_CHR_DEV_OSST)	+= osst.o
 obj-$(CONFIG_BLK_DEV_SD)	+= sd.o
 obj-$(CONFIG_BLK_DEV_SR)	+= sr_mod.o
 obj-$(CONFIG_CHR_DEV_SG)	+= sg.o
diff -ur linux-2.4.0-orig/drivers/scsi/hosts.c linux-2.4.0-test8/drivers/scsi/hosts.c
--- linux-2.4.0-orig/drivers/scsi/hosts.c	Tue Jul 11 14:17:45 2000
+++ linux-2.4.0-test8/drivers/scsi/hosts.c	Sun Sep 17 10:01:49 2000
@@ -1008,6 +1008,9 @@
 #ifdef CONFIG_CHR_DEV_ST
     scsi_register_device(&st_template);
 #endif
+#ifdef CONFIG_CHR_DEV_OSST
+    scsi_register_device(&osst_template);
+#endif
 #ifdef CONFIG_CHR_DEV_SG
     scsi_register_device(&sg_template);
 #endif
diff -ur linux-2.4.0-orig/drivers/scsi/osst_detect.h linux-2.4.0-test8/drivers/scsi/osst_detect.h
--- linux-2.4.0-orig/drivers/scsi/osst_detect.h	Sun Sep 17 10:04:59 2000
+++ linux-2.4.0-test8/drivers/scsi/osst_detect.h	Sun Sep 17 15:53:38 2000
@@ -0,0 +1,6 @@
+#define SIGS_FROM_OSST \
+       {"OnStream", "SC-", "", "osst"}, \
+       {"OnStream", "DI-", "", "osst"}, \
+       {"OnStream", "DP-", "", "osst"}, \
+       {"OnStream", "FW-", "", "osst"}, \
+       {"OnStream", "USB", "", "osst"}
diff -ur linux-2.4.0-orig/drivers/scsi/st.c linux-2.4.0-test8/drivers/scsi/st.c
--- linux-2.4.0-orig/drivers/scsi/st.c	Sun Sep 17 09:47:10 2000
+++ linux-2.4.0-test8/drivers/scsi/st.c	Sun Sep 17 15:55:27 2000
@@ -175,7 +175,7 @@
 static int st_int_ioctl(Scsi_Tape *, unsigned int, unsigned long);
 
 
-/* #include "osst_detect.h" */
+#include "osst_detect.h"
 #ifndef SIGS_FROM_OSST
 #define SIGS_FROM_OSST \
 	{"OnStream", "SC-", "", "osst"}, \
diff -ur linux-2.4.0-orig/include/linux/blk.h linux-2.4.0-test8/include/linux/blk.h
--- linux-2.4.0-orig/include/linux/blk.h	Wed Aug  9 21:59:31 2000
+++ linux-2.4.0-test8/include/linux/blk.h	Sun Sep 17 16:04:33 2000
@@ -164,6 +164,14 @@
 #define DEVICE_INTR do_st  
 #define DEVICE_NR(device) (MINOR(device) & 0x7f)
 
+#elif (MAJOR_NR == OSST_MAJOR)
+
+#define DEVICE_NAME "onstream" 
+#define DEVICE_INTR do_osst
+#define DEVICE_NR(device) (MINOR(device) & 0x7f) 
+#define DEVICE_ON(device) 
+#define DEVICE_OFF(device) 
+
 #elif (MAJOR_NR == SCSI_CDROM_MAJOR)
 
 #define DEVICE_NAME "CD-ROM"
@@ -329,7 +337,7 @@
 #define DEVICE_OFF(device) do {} while (0)
 #endif
 
-#if (MAJOR_NR != SCSI_TAPE_MAJOR)
+#if (MAJOR_NR != SCSI_TAPE_MAJOR) && (MAJOR_NR != OSST_MAJOR)
 #if !defined(IDE_DRIVER)
 
 #ifndef CURRENT
diff -ur linux-2.4.0-orig/include/linux/major.h linux-2.4.0-test8/include/linux/major.h
--- linux-2.4.0-orig/include/linux/major.h	Thu Sep 14 22:39:31 2000
+++ linux-2.4.0-test8/include/linux/major.h	Sun Sep 17 10:01:49 2000
@@ -143,6 +143,8 @@
 #define MSR_MAJOR		202
 #define CPUID_MAJOR		203
 
+#define OSST_MAJOR	206	/* OnStream-SCx0 SCSI tape */
+
 /*
  * Tests for SCSI devices.
  */
