diff -uNr linux-2.4.0-test3.fixed/MAINTAINERS linux-2.4.0-test3.osst/MAINTAINERS
--- linux-2.4.0-test3.fixed/MAINTAINERS	Sun Jul  9 19:10:44 2000
+++ linux-2.4.0-test3.osst/MAINTAINERS	Sat Jul 15 16:07:14 2000
@@ -798,6 +798,13 @@
 W:	http://www.linuxtr.net
 S:	Maintained
 
+ONSTREAM SCSI TAPE DRIVER
+P:	Willem Riede
+M:	wriede@monmouth.com
+L:	osst@linux1.onstream.nl
+L:	linux-scsi@vger.rutgers.edu
+S:	Maintained
+
 OPL3-SA2, SA3, and SAx DRIVER
 P:	Scott Murray
 M:	scott@spiteful.org
diff -uNr linux-2.4.0-test3.fixed/drivers/scsi/Config.in linux-2.4.0-test3.osst/drivers/scsi/Config.in
--- linux-2.4.0-test3.fixed/drivers/scsi/Config.in	Mon Mar 27 19:45:33 2000
+++ linux-2.4.0-test3.osst/drivers/scsi/Config.in	Sat Jul 15 16:07:14 2000
@@ -8,6 +8,8 @@
 
 dep_tristate '  SCSI tape support' CONFIG_CHR_DEV_ST $CONFIG_SCSI
 
+dep_tristate '  OnStream SC-x0 SCSI tape support' CONFIG_CHR_DEV_OSST $CONFIG_SCSI
+
 dep_tristate '  SCSI CD-ROM support' CONFIG_BLK_DEV_SR $CONFIG_SCSI
 
 if [ "$CONFIG_BLK_DEV_SR" != "n" ]; then
diff -uNr linux-2.4.0-test3.fixed/drivers/scsi/Makefile linux-2.4.0-test3.osst/drivers/scsi/Makefile
--- linux-2.4.0-test3.fixed/drivers/scsi/Makefile	Tue Jun 20 23:14:51 2000
+++ linux-2.4.0-test3.osst/drivers/scsi/Makefile	Sat Jul 15 16:07:14 2000
@@ -31,6 +31,7 @@
 
 obj-$(CONFIG_SCSI)		+= scsi_mod.o
 obj-$(CONFIG_CHR_DEV_ST)	+= st.o
+obj-$(CONFIG_CHR_DEV_OSST)	+= osst.o
 obj-$(CONFIG_BLK_DEV_SD)	+= sd.o
 obj-$(CONFIG_BLK_DEV_SR)	+= sr_mod.o
 obj-$(CONFIG_CHR_DEV_SG)	+= sg.o
diff -uNr linux-2.4.0-test3.fixed/drivers/scsi/hosts.c linux-2.4.0-test3.osst/drivers/scsi/hosts.c
--- linux-2.4.0-test3.fixed/drivers/scsi/hosts.c	Mon Jun 26 20:40:38 2000
+++ linux-2.4.0-test3.osst/drivers/scsi/hosts.c	Sat Jul 15 16:07:14 2000
@@ -998,6 +998,9 @@
 #ifdef CONFIG_CHR_DEV_ST
     scsi_register_device(&st_template);
 #endif
+#ifdef CONFIG_CHR_DEV_OSST
+    scsi_register_device(&osst_template);
+#endif
 #ifdef CONFIG_CHR_DEV_SG
     scsi_register_device(&sg_template);
 #endif
diff -uNr linux-2.4.0-test3.fixed/drivers/scsi/hosts.h linux-2.4.0-test3.osst/drivers/scsi/hosts.h
--- linux-2.4.0-test3.fixed/drivers/scsi/hosts.h	Wed Jul 12 13:10:54 2000
+++ linux-2.4.0-test3.osst/drivers/scsi/hosts.h	Sat Jul 15 16:07:14 2000
@@ -508,6 +508,7 @@
 
 extern struct Scsi_Device_Template sd_template;
 extern struct Scsi_Device_Template st_template;
+extern struct Scsi_Device_Template osst_template;
 extern struct Scsi_Device_Template sr_template;
 extern struct Scsi_Device_Template sg_template;
 
diff -uNr linux-2.4.0-test3.fixed/drivers/scsi/osst_detect.h linux-2.4.0-test3.osst/drivers/scsi/osst_detect.h
--- linux-2.4.0-test3.fixed/drivers/scsi/osst_detect.h	Thu Jan  1 01:00:00 1970
+++ linux-2.4.0-test3.osst/drivers/scsi/osst_detect.h	Sat Jul 15 16:07:14 2000
@@ -0,0 +1,5 @@
+#define OSST_SUPPORTS(SDp) (! ( memcmp (SDp->vendor, "OnStream", 8) || \
+			     ( memcmp (SDp->model, "SC-", 3) && \
+			       memcmp (SDp->model, "DI-", 3) && \
+			       memcmp (SDp->model, "DP-", 3) && \
+			       memcmp (SDp->model, "USB", 3) ) ) )
diff -uNr linux-2.4.0-test3.fixed/drivers/scsi/st.c linux-2.4.0-test3.osst/drivers/scsi/st.c
--- linux-2.4.0-test3.fixed/drivers/scsi/st.c	Mon Jun 26 21:33:03 2000
+++ linux-2.4.0-test3.osst/drivers/scsi/st.c	Sat Jul 15 16:07:14 2000
@@ -65,6 +65,8 @@
 
 #include "constants.h"
 
+#include "osst_detect.h"
+
 static int buffer_kbs;
 static int write_threshold_kbs;
 static int max_buffers = (-1);
@@ -3426,6 +3428,18 @@
 #endif
 
 
+/* Returns zero for drives not supported by this driver */
+static int st_supported(Scsi_Device * SDp)
+{
+  /* These OnStream SC drives require a special driver. The OnStream ADR* drives
+     are supported by this driver */
+  if ( OSST_SUPPORTS(SDp) )
+      return 0;
+
+  return 1;
+}
+
+
 static struct file_operations st_fops =
 {
 	read:		st_read,
@@ -3446,6 +3460,8 @@
 
 	if (SDp->type != TYPE_TAPE)
 		return 1;
+	if (!st_supported(SDp))
+		return 1;
 
 	write_lock_irqsave(&st_dev_arr_lock, flags);
 	if (st_template.nr_dev >= st_template.dev_max) {
@@ -3610,6 +3626,8 @@
 static int st_detect(Scsi_Device * SDp)
 {
 	if (SDp->type != TYPE_TAPE)
+		return 0;
+	if (!st_supported(SDp))
 		return 0;
 
 	printk(KERN_WARNING
diff -uNr linux-2.4.0-test3.fixed/include/linux/blk.h linux-2.4.0-test3.osst/include/linux/blk.h
--- linux-2.4.0-test3.fixed/include/linux/blk.h	Wed Jul 12 12:53:48 2000
+++ linux-2.4.0-test3.osst/include/linux/blk.h	Sat Jul 15 16:07:14 2000
@@ -176,6 +176,14 @@
 #define DEVICE_ON(device)
 #define DEVICE_OFF(device)
 
+#elif (MAJOR_NR == OSST_MAJOR)
+
+#define DEVICE_NAME "onstream" 
+#define DEVICE_INTR do_osst
+#define DEVICE_NR(device) (MINOR(device) & 0x7f) 
+#define DEVICE_ON(device) 
+#define DEVICE_OFF(device) 
+
 #elif (MAJOR_NR == SCSI_CDROM_MAJOR)
 
 #define DEVICE_NAME "CD-ROM"
@@ -383,7 +391,7 @@
 
 #endif /* MAJOR_NR == whatever */
 
-#if (MAJOR_NR != SCSI_TAPE_MAJOR)
+#if (MAJOR_NR != SCSI_TAPE_MAJOR) && (MAJOR_NR != OSST_MAJOR)
 #if !defined(IDE_DRIVER)
 
 #ifndef CURRENT
diff -uNr linux-2.4.0-test3.fixed/include/linux/major.h linux-2.4.0-test3.osst/include/linux/major.h
--- linux-2.4.0-test3.fixed/include/linux/major.h	Mon Jun 19 22:42:43 2000
+++ linux-2.4.0-test3.osst/include/linux/major.h	Sat Jul 15 16:07:14 2000
@@ -141,6 +141,8 @@
 #define MSR_MAJOR		202
 #define CPUID_MAJOR		203
 
+#define OSST_MAJOR	206	/* OnStream-SCx0 SCSI tape */
+
 /*
  * Tests for SCSI devices.
  */
diff -uNr linux-2.4.0-test3.fixed/include/linux/mtio.h linux-2.4.0-test3.osst/include/linux/mtio.h
--- linux-2.4.0-test3.fixed/include/linux/mtio.h	Wed Jul 12 12:54:44 2000
+++ linux-2.4.0-test3.osst/include/linux/mtio.h	Sat Jul 15 16:07:14 2000
@@ -105,6 +105,8 @@
 #define MT_ISEVEREX_FT40A	0x32	/* Everex FT40A (QIC-40) */
 #define MT_ISDDS1		0x51	/* DDS device without partitions */
 #define MT_ISDDS2		0x52	/* DDS device with partitions */
+#define MT_ISONSTREAM_SC        0x61	/* OnStream SCSI tape drives (SC-x0)
+					   and SCSI emulated (DI, DP, USB) */
 #define MT_ISSCSI1		0x71	/* Generic ANSI SCSI-1 tape unit */
 #define MT_ISSCSI2		0x72	/* Generic ANSI SCSI-2 tape unit */
 
@@ -134,6 +136,7 @@
 	{MT_ISWT5099EEN24,	"Wangtek 5099-een24, 60MB"}, \
 	{MT_ISTEAC_MT2ST,	"Teac MT-2ST 155mb data cassette drive"}, \
 	{MT_ISEVEREX_FT40A,	"Everex FT40A, QIC-40"}, \
+	{MT_ISONSTREAM_SC,      "OnStream SC-, DI-, DP-, or USB tape drive"}, \
 	{MT_ISSCSI1,		"Generic SCSI-1 tape"}, \
 	{MT_ISSCSI2,		"Generic SCSI-2 tape"}, \
 	{0, NULL} \
